### DISCUSSION TITLE
Validators don't get triggered for  Ref array

### DISCUSSION THREAD

**Discussion by dor272** [typegoose/typegoose#530] [Q&A] [answered]
Hello,

I have a model referencing another model in a property which is array.
I have tried to add custom validator on the prop options, innerOptions and outerOptions in order to trigger a validation - none worked.

The prebuilt validators didn't work either (such as `minLength`).

Here is a test I ran:
```ts
import {
    prop,
    Ref,
    getModelForClass,
} from '@typegoose/typegoose';

class Y {
    @prop({ required: true })
    name!: string
}

class X {
    @prop({ required: true })
    name!: string

    @prop({
        ref: 'Y',
        required: true,
        validate: {
            validator: (v: any) => {
                console.log(v, 'v');
                return false;
            },
        },
        outerOptions: {
            validator: (v: any) => {
                console.log(v, 'ov');
                return false;
            },
        },
        innerOptions: {
            validator: (v: any) => {
                console.log(v, 'iv');
                return false;
            },
        },
    })
    yArray!: Ref<Y>[]
}

const XModel = getModelForClass(X);
async function test() {
    const x = new XModel({
        name: 'test',
        yArray: [],
    });
    await x.validate();
    console.log('passed validation');
}

test();
```

I'm using the following versions: 
node: 14.15.1
typegoose: 7.4.8
mongoose: 5.10.18
@types/mongoose: 5.10.5 

I don't know  if I'm doing something wrong or  it's a bug.

Would appreciate any help,

Thanks

**Comment by hasezoey** [comment, marked as answer]
depending on what you want to validate, you could try to move the validation to the referenced model

could you maybe try the typegoose beta (currently `8.0.0-beta.5` with mongoose `5.12.4` (maybe higher works too)) to see if there is the same behaviour? (on properties with `ref`, the options `innerOptions` and `outerOptions` do nothing before 8.0.0)

**Reply by dor272** [reply]
I want to validate for example the array length or to check wether there's a duplicate elements in it, that's irrelevant to the referenced model.

I've ran some tests with the versions you suggested:

Now the validate did get triggered (both when providing single or multiple validator functions), but a prebuilt validator didn't work:

```ts
import {
    prop,
    Ref,
    getModelForClass,
} from '@typegoose/typegoose';

class Y {
    @prop({ required: true })
    name!: string
}

class X {
    @prop({ required: true })
    name!: string

    @prop({
        ref: 'Y',
        required: true,
        validate: {
            validator: (v: any) => {
                return v.length > 0;
            },
        },
       })
    yArray!: Ref<Y>[]
}

const XModel = getModelForClass(X);
async function test() {
    const x = new XModel({
        name: 'test',
        yArray: [],
    });
    await x.validate();
    console.log('passed validation');
}

test();
```
worked (failed as expected), but changing it to:

```ts
class X {
...
    @prop({
        ref: 'Y',
        required: true,
        minLength: 1
    })
   ...
}
```
failed (passed validation).

**Reply by hasezoey** [reply]
if i understand correctly your first code example works as expected, but your second code example (`minLength`) does not work as expected?

when yes, this is because `minlength` is an validator for *Strings* not *Arrays* (and for numbers its `min`) arrays dont have an built-in validator
